dali/operators/generic/slice/slice_base.cu

```diff
@@ -13,82 +13,99 @@
-    slice_args.reserve(slice_anchors.size());
-    for (std::size_t i = 0; i < slice_anchors.size(); i++) {
-      std::array<int64_t, NumDims> anchor, shape;
-      const auto& slice_anchor = slice_anchors[i];
-      const auto& slice_shape = slice_shapes[i];
-      for (std::size_t d = 0; d < NumDims; d++) {
-        anchor[d] = slice_anchor[d];
-        shape[d] = slice_shape[d];
-      }
-      slice_args.push_back({anchor, shape});
-    }
 
-    kernels::KernelRequirements req = kernel.Setup(ctx, in_view, slice_args);
+template <typename OutputType, typename InputType, int Dims>
+class SliceBaseGpu : public OpImplBase<GPUBackend> {
+ public:
+  using Kernel = kernels::SliceGPU<OutputType, InputType, Dims>;
+  using SliceArgs = kernels::SliceArgs<OutputType, Dims>;
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<GPUBackend> &ws) override;
+  void RunImpl(workspace_t<GPUBackend> &ws) override;
+
+  std::vector<SliceArgs>& Args() { return args_; }
+
+ private:
+  std::vector<SliceArgs> args_;
+  kernels::KernelManager kmgr_;
+};
+
+template <typename OutputType, typename InputType, int Dims>
+bool SliceBaseGpu<OutputType, InputType, Dims>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                                          const workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.template InputRef<GPUBackend>(0);
+  auto in_shape = input.shape();
+  int nsamples = in_shape.num_samples();
 
-    output.set_type(TypeInfo::Create<OutputType>());
-    output.Resize(req.output_shapes[0]);
+  output_desc.resize(1);
+  output_desc[0].type = TypeInfo::Create<OutputType>();
+  output_desc[0].shape.resize(nsamples, Dims);
 
-    scratch_alloc.Reserve(req.scratch_sizes);
-    auto scratchpad = scratch_alloc.GetScratchpad();
-    ctx.scratchpad = &scratchpad;
+  kmgr_.Resize<Kernel>(1, 1);
+  auto in_view = view<const InputType, Dims>(input);
+  kernels::KernelContext ctx;
+  auto req = kmgr_.Setup<Kernel>(0, ctx, in_view, args_);
+  output_desc[0].shape = req.output_shapes[0];
+  return true;
+}
+
+template <typename OutputType, typename InputType, int Dims>
+void SliceBaseGpu<OutputType, InputType, Dims>::RunImpl(workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.template InputRef<GPUBackend>(0);
+  auto &output = ws.template OutputRef<GPUBackend>(0);
 
-    auto out_view = view<OutputType, NumDims>(output);
-    kernel.Run(ctx, out_view, in_view, slice_args);
-  ),  // NOLINT
-  (
-    DALI_FAIL("Not supported number of dimensions: " + std::to_string(number_of_dims));
-  ));  // NOLINT
+  auto in_view = view<const InputType, Dims>(input);
+  auto out_view = view<OutputType, Dims>(output);
+  kernels::KernelContext ctx;
+  ctx.gpu.stream = ws.stream();
+  kmgr_.Run<Kernel>(0, 0, ctx, out_view, in_view, args_);
+  output.SetLayout(input.GetLayout());
 }
 
-}  // namespace detail
+template <>
+bool SliceBase<GPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
```

Comment:
Isn't this the same as for CPU? (Same with RunImpl).

Reply:
Yes, pretty much, except for the usage of GPU and CPU kernel respective. I don't want to include both kernels in the same header